Adds "All in the Cards" repeatable quest to Chululu

Era-Lusiphur opened the pull request:
Adds "All in the Cards" repeatable quest to Chululu to help increase server supply of Tarut Cards used in the "Collect Taru Cards" pre-req quest for Sleepga II quest line."

Currently setup to require "All in the Cards" and "Rubbish Day" to be completed first.

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [X] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [X] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Era-Lusiphur committed:
Adds "All in the Cards" repeatable quest to Chululu

Adds "All in the Cards" repeatable quest to Chululu to help increase server supply of Tarut Cards used in the "Collect Taru Cards" pre-req quest for Sleepga II quest line.

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -19,12 +19,20 @@ function onTrade(player,npc,trade)
             player:startEvent(200); -- Finish quest "Collect Tarut Cards"
         end
     end
+    if (player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS) == QUEST_ACCEPTED) then
```
Should probably check `>= QUEST_ACCEPTED` since this is repeatable.

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -98,7 +117,82 @@ function onEventFinish(player,csid,option)
+            player:messageSpecial(ID.text.ITEM_OBTAINED,card);
+        end	
+	elseif (csid == 10113 and option==0) then  -- ALL_IN_THE_CARDS repeat started
+        player:delQuest(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS);	
```
Does this actually get removed from your completed quests if you start it again?

Era-Lusiphur replied:
Yes, I believe it does. Was the only method I knew of allowing it to be repeatable. Rookie things, right?

TeoTwawki replied:
repeatables normally don't delete. they stay in the completed list till activated again and jump straight to active.

Era-Lusiphur replied:
The More You Know™, the less ghetto your PRs are. :P

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -98,7 +117,82 @@ function onEventFinish(player,csid,option)
+        player:messageSpecial(ID.text.GIL_OBTAINED,GIL_RATE*600);
+		player:setCharVar("AllInTheCards_date",0);		
+        player:tradeComplete();	
+        player:completeQuest(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS);		
```
```lua
if npcUtil.completeQuest(player, JEUNO, ALL_IN_THE_CARDS, {
        gil = 600,
        title = tpz.title.CARD_COLLECTOR,
        var = {"AllInTheCards_date"} })
then
    trade:confirm()
```

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -19,12 +19,20 @@ function onTrade(player,npc,trade)
         end
     end
+    if (player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS) == QUEST_ACCEPTED) then
+        if (trade:hasItemQty(558,1) == true and trade:hasItemQty(559,1) == true and trade:hasItemQty(561,1) == true and trade:hasItemQty(562,1) == true and trade:getItemCount() == 4) then
+            player:startEvent(10114); -- Finish quest "All in the Cards"
+        end
```
```lua
if npcUtil.tradeHas(trade, {558, 559, 561, 562}, true) then
    player:startEvent(10114) -- Finish quest "All in the Cards"
end
````

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -98,7 +117,82 @@ function onEventFinish(player,csid,option)
         player:setCharVar("RubbishDay_prog",0);
-        player:setCharVar("RubbishDay_day",0);
+        player:setCharVar("RubbishDay_day",VanadielDayOfTheYear());
+	elseif (csid == 10110 and option==0) then  -- ALL_IN_THE_CARDS started
```
CSIDs 10110, 10113, and 10112 can probably all put into the same block
```lua
if (csid == 10110 or csid == 10112 or csid == 10113) and option == 0 then
--or--
if csid == 10112 or ((csid == 10110 or csid == 10113) and option == 0) then
```

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -19,12 +19,20 @@ function onTrade(player,npc,trade)
     local RubbishDay = player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.RUBBISH_DAY);
     local SearchingForTheRightWords = player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.SEARCHING_FOR_THE_RIGHT_WORDS);
+    local AllInTheCards = player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS);
+    local realday = tonumber(os.date("%j")); -- %M for next minute, %j for next day	
+    local cdate = player:getCharVar("AllInTheCards_date");
```
Change to `getMidnight()` as the charvar then you can get rid of `realday`.

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -51,6 +59,17 @@ function onTrigger(player,npc)
+
+    elseif (player:getFameLevel(JEUNO) >= 4 and CollectTarutCards == QUEST_COMPLETED and AllInTheCards == QUEST_AVAILABLE) then
+        player:startEvent(10110); -- Start quest "All in the Cards" with option
+	elseif (AllInTheCards == QUEST_ACCEPTED and cdate == realday) then
+		player:startEvent(10111); -- During quest "All in the Cards" and same AllInTheCards_date value		
+	elseif (AllInTheCards == QUEST_ACCEPTED and cdate ~= realday) then
```
If you use `getMidnight()` as the charvar, you can just do this:
```lua
os.time() < cdate -- need to wait till jp midnight
os.time() >= cdate -- good to complete
```

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -98,7 +117,82 @@ function onEventFinish(player,csid,option)
+            card = 558; -- Tarut: Fool
+        end
+
+        if (player:getFreeSlotsCount() == 0) then
+            player:messageSpecial(ID.text.ITEM_CANNOT_BE_OBTAINED,card);
+        else	
+            player:addQuest(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS);
+			player:setCharVar("AllInTheCards_date", os.date("%j")); -- %M for next minute, %j for next day	
+            player:addItem(card,5);
+            player:messageSpecial(ID.text.ITEM_OBTAINED,card);
+        end	
```
```lua
if npcUtil.giveItem(player, {{card, 5}}) then
    player:setCharVar("AllInTheCards_date", getMidnight())
end
```

Era-Lusiphur committed:
Made requested changes, tested

Made requested changes, tested

Era-Lusiphur commented:
Made requested changes, tested all of her quests. Seems good to go here.

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -43,16 +52,30 @@ function onTrigger(player,npc)
+		
+    elseif (player:getFameLevel(JEUNO) >= 4 and CollectTarutCards == QUEST_COMPLETED and AllInTheCards == QUEST_AVAILABLE) then
+        player:startEvent(10110); -- Start quest "All in the Cards" with option
+	elseif (AllInTheCards == QUEST_ACCEPTED and cdate >= os.time()) then
+		player:startEvent(10111); -- During quest "All in the Cards" and same AllInTheCards_date value		
+	elseif (AllInTheCards == QUEST_ACCEPTED and cdate ~= os.time()) then
+		player:startEvent(10112); -- During quest "All in the Cards"  THIS ONE GIVES ANOTHER BATCH
+	elseif (CollectTarutCards == QUEST_COMPLETED and AllInTheCards == QUEST_COMPLETED) then
+		player:startEvent(10113); -- Start quest "All in the Cards"	repeat with option		
```
This will need to check that AllInTheCards >= QUEST_ACCEPTED for people repeating the quest. Something like:
```lua
elseif AllInTheCards >= QUEST_ACCEPTED then
    if cdate >= os.time() then
        player:startEvent(10111) 
    elseif cdate == 0 then
        player:startEvent(10113)
    elseif cdate < os.time() then
        player:startEvent(10112) 
    end
```
Once a quest is marked as complete it stays that way unless it's manually deleted. Adding the quest again doesn't set it back to accepted. Once you complete the quest, it will always give you cs 10113 and never continue to the stuff past this block. I'm not sure how this quest is supposed to progress, but you will probably need to add a localvar to track whether the player has denied this quest/timer isn't up yet so that they can talk again and get the next quest dialogue.

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -98,18 +121,43 @@ function onEventFinish(player,csid,option)
-            player:messageSpecial(ID.text.GIL_OBTAINED,GIL_RATE*6000);
-            player:addItem(13083);
-            player:messageSpecial(ID.text.ITEM_OBTAINED,13083);
-            player:setCharVar("RubbishDayVar",0);
-            player:addFame(JEUNO, 30);
-            player:completeQuest(JEUNO,tpz.quest.id.jeuno.RUBBISH_DAY);
+			npcUtil.completeQuest(player, JEUNO, tpz.quest.id.jeuno.RUBBISH_DAY, {
+			gil = 6000,
+			item = 13083,
+			var = {"RubbishDayVar"} })
```
You can get rid of this:
```lua
        if (player:getFreeSlotsCount() == 0) then
            player:messageSpecial(ID.text.ITEM_CANNOT_BE_OBTAINED,13083);
        else
```
and just have the npcUtil.completeQuest.

cocosolos reviewed (changes requested):
Few more things from me, also a lot of extra spaces/tabs/lines that should be removed. Check the [style guide](https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#style-guide). No tabs, no semi-colons, no ( ) if you can help it. e.g. 
```lua
if (player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS) >= QUEST_ACCEPTED) then
```
should be 
```lua
if player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS) >= QUEST_ACCEPTED then
```

Era-Lusiphur commented:
> Few more things from me, also a lot of extra spaces/tabs/lines that should be removed. Check the [style guide](https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#style-guide). 

Sorry about that, wasn't aware of it until Teo mentioned it the other day to me in a PM. I will get this adjusted soon and resubmitted.

zircon-tpl commented:
Hello, @Era-Lusiphur !

I am ensuring that this Pull Request has not been forgotten. Is there anything you might need assistance with?

Era-Lusiphur commented:
> I am ensuring that this Pull Request has not been forgotten. Is there anything you might need assistance with?

Just been busy with other things. I apologize about the delay.

zircon-tpl commented:
That is perfectly acceptable! Please do not hesitate to let us know if we may assist in some manner. Project Topaz strives to work with Contributors!

Era-Lusiphur committed:
Update Chululu.lua

Made changes to conform to style guide and additional changes discussed with cocosolos.

TeoTwawki commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -1,115 +1,149 @@
         end
     end
-end;
+    if player:getQuestStatus(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS) >= QUEST_ACCEPTED then
```
Remove the "end" right before here and use `elseif` - the only reason to use two separate if/end blocks is if both things can execute, which here that is not the case.

TeoTwawki commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -1,115 +1,149 @@
+		if npcUtil.giveItem(player, {{card, 5}}) then
+            player:addQuest(JEUNO,tpz.quest.id.jeuno.ALL_IN_THE_CARDS)
+		player:setCharVar("AllInTheCards_date", getMidnight())
+        player:setLocalVar("Cardstemp", 1)
```
indentation went a little odd around here due to tabs.

Era-Lusiphur replied:
Weird how that got formatted. Silly me, assuming EditorConfig took care of it all. On my end, all lined up.

ibm2431 replied:
May be some leftover tabs in the file, and it's lining up in your editor because the "indent space" for a tab is set to be equal width to four actual spaces. End result is it looks okay on your screen, but if that tab space amount is different on a different display (ie: here on GitHub) it'll no longer line up.

TeoTwawki replied:
> Silly me, assuming EditorConfig took care of it all. 

it actually should have. dunno why it didn't.

Era-Lusiphur committed:
Update Chululu.lua

More style changes, corrected end/elseif

Era-Lusiphur committed:
Update Chululu.lua

Updated initial trade for prior quest to use npcUtil.tradeHas and trim fat

cocosolos commented on scripts/zones/Lower_Jeuno/npcs/Chululu.lua:
```diff
@@ -1,115 +1,148 @@
+        player:startEvent(10110) -- Start quest "All in the Cards" with option
+    elseif AllInTheCards >= QUEST_ACCEPTED and player:getLocalVar("Cardstemp") == 0 then
+        if cdate >= os.time() then
+            player:startEvent(10111) -- During quest "All in the Cards" and same AllInTheCards_date value
```
I think the only other issue I see is there is nothing in onEventFinish that sets `Cardstemp` for this CS, so after they accept the quest but before they complete it, while they're waiting, the other quests aren't accessible.

Era-Lusiphur replied:
Change made,  coco

Era-Lusiphur committed:
Update Chululu.lua

adds localvar to cs for same day so that other quest CS, dialogue can be triggered
